streams/src/test/java/org/apache/kafka/streams/integration/PositionCheckpointIntegrationTest.java

```diff
@@ -0,0 +1,796 @@
+        } else if (Objects.equals(kind, "DSL") && supplier instanceof WindowBytesStoreSupplier) {
+            setUpWindowDSLTopology((WindowBytesStoreSupplier) supplier, builder);
+        } else if (Objects.equals(kind, "PAPI") && supplier instanceof WindowBytesStoreSupplier) {
+            setUpWindowPAPITopology((WindowBytesStoreSupplier) supplier, builder);
+        } else if (Objects.equals(kind, "DSL") && supplier instanceof SessionBytesStoreSupplier) {
+            setUpSessionDSLTopology((SessionBytesStoreSupplier) supplier, builder);
+        } else if (Objects.equals(kind, "PAPI") && supplier instanceof SessionBytesStoreSupplier) {
+            setUpSessionPAPITopology((SessionBytesStoreSupplier) supplier, builder);
+        } else {
+            throw new AssertionError("Store supplier is an unrecognized type.");
+        }
+
+        kafkaStreams =
+                IntegrationTestUtils.getStartedStreams(
+                        streamsConfig,
+                        builder,
+                        cleanup
+                );
+    }
+
+    @After
+    public void afterTest() {
+        afterTest(true);
+    }
+
+    public void afterTest(final boolean cleanup) {
+        // only needed because some of the PAPI cases aren't added yet.
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+            if (cleanup) {
+                kafkaStreams.cleanUp();
+            }
+        }
+    }
+
+    @AfterClass
+    public static void after() {
+        CLUSTER.stop();
+    }
+
+    public void reboot() {
+        afterTest(false);
+        beforeTest(false);
+    }
+
+    @Test
+    public void verifyStore() throws Exception {
+        if (storeToTest.keyValue()) {
+            shouldCheckpointKeyValueStore(2);
+        }
+
+        if (storeToTest.isWindowed()) {
+            shouldHandleWindowKeyQueries();
+        }
+
+        if (storeToTest.isSession()) {
+            shouldHandleSessionKeyQueries();
+        }
+    }
+    public <V> void shouldCheckpointKeyValueStore(
+            final Integer key) throws Exception {
+        shouldHandleKeyQuery(key);
+        reboot();
+        shouldHandleKeyQuery(key);
+    }
+
+    private <T> void shouldHandleWindowKeyQueries() {
+        shouldHandleWindowKeyQuery(
+            2,
+            Instant.ofEpochMilli(WINDOW_START),
+            Instant.ofEpochMilli(WINDOW_START)
+        );
+        reboot();
+        shouldHandleWindowKeyQuery(
+                2,
+                Instant.ofEpochMilli(WINDOW_START),
+                Instant.ofEpochMilli(WINDOW_START)
+        );
+    }
+
+    private <T> void shouldHandleSessionKeyQueries() {
+        shouldHandleSessionRangeQuery(
+            2,
+            mkSet(2)
+        );
+        reboot();
+        shouldHandleSessionRangeQuery(
+                2,
+                mkSet(2)
+        );
+    }
+
+    public <V> void shouldHandleKeyQuery(
```

Comment:
What we want to test in these tests here is that the query returns the same results (same rows) and that the position is the same. It is not enough to check that the query has not failed